### DISCUSSION TITLE
Improve built-in support for commonly used messages

### DISCUSSION THREAD

**Discussion by jtbandes** [foxglove/community#168] [Q&A]
Tracking some well-known message types which we could consider adding built-in panels/visualizations for.

- https://docs.ros.org/en/melodic/api/sensor_msgs/html/index-msg.html 
- https://github.com/ros-planning/moveit_msgs
- https://github.com/ros-perception/vision_msgs
- https://github.com/luxonis/depthai-ros/tree/noetic-devel/depthai_ros_msgs/msg
- http://wiki.ros.org/shape_msgs
- https://github.com/ANYbotics/grid_map/blob/master/grid_map_msgs/msg/GridMap.msg
- https://github.com/uos/mesh_tools / https://www.researchgate.net/publication/336627583_Tools_for_Visualizing_Annotating_and_Storing_Triangle_Meshes_in_ROS_and_RViz
- https://github.com/OctoMap/octomap_msgs
- Wrench: http://wiki.ros.org/rviz/DisplayTypes/Wrench

- foxglove/studio#782 
- foxglove/community#156
- foxglove/community#146
- foxglove/studio#1630
- foxglove/studio#1972
- foxglove/studio#1325
- foxglove/community#155
- foxglove/community#113
- foxglove/studio#1049
- foxglove/studio#1615
- foxglove/studio#2812
- foxglove/studio#2895
- foxglove/studio#3156
- foxglove/studio#3213
- foxglove/studio#3215
- foxglove/studio#3088
- https://github.com/foxglove/studio/issues/2359
- https://github.com/foxglove/studio/issues/3396
- https://github.com/foxglove/studio/issues/3932
- https://github.com/foxglove/studio/issues/3315

**Comment by wkalt** [comment]
the tutorial here also allows easy interaction with the robot navigation in rviz, by clicking a location on the map:

https://emanual.robotis.com/docs/en/platform/turtlebot3/quick-start/

**Comment by adamconkey** [comment]
+1 for `moveit_msgs`, in particular the `DisplayRobotState` and `DisplayTrajectory` are incredibly useful for previewing trajectories and visualizing key waypoints in a the trajectory of a robot arm. Is it recommended to create a separate issue for more specific feature requests?

**Comment by jtbandes** [comment]
Yes I suppose we could stand to have a separate ticket for the moveit stuff. Is there a reference implementation with screenshots demonstrating how the types are meant to be visualized?

**Comment by adamconkey** [comment]
It looks like the visualization code is in [rviz_plugin_render_tools](https://github.com/ros-planning/moveit2/tree/132ad2853960c4cb9da47bdd49dab8ebad2dd723/moveit_ros/visualization/rviz_plugin_render_tools) and in particular there is [trajectory_visualization.cpp](https://github.com/ros-planning/moveit2/blob/132ad2853960c4cb9da47bdd49dab8ebad2dd723/moveit_ros/visualization/rviz_plugin_render_tools/src/trajectory_visualization.cpp). There are some examples in the [videos/images of the MoveIt tutorials](https://moveit.picknik.ai/humble/doc/tutorials/quickstart_in_rviz/quickstart_in_rviz_tutorial.html) for the trajectory visualization. It looks similar to the URDF visualization in that you're showing the robot with its meshes in the scene, but it's decoupled from a published robot state so that you can e.g. preview a trajectory before executing it, or visualize multiple different plans at once. They also support coloring robot links selectively, which is useful e.g. color links that will collide with an object differently than those that are free of collision.

**Comment by vinnnyr** [comment]
Wanted to link my request to visualize PointStamped to this issue as well 
https://github.com/orgs/foxglove/discussions/579

**Comment by rg-gravis** [comment]
Seconding a request for [Grid Map](https://github.com/ANYbotics/grid_map/) support (not sure if it belongs here or in https://github.com/orgs/foxglove/discussions/433), I was able to write a Message Converter that allows for basic visualization (and would be happy to share the code if that's helpful) but am blocked from implementing the full featureset of the existing [grid_map_rviz_plugin](https://github.com/ANYbotics/grid_map/tree/master?tab=readme-ov-file#grid_map_rviz_plugin) in Foxglove because there's no way to get user input into the message converter. The most important feature of that plugin that is impossible to do in Foxglove is the ability to (independently, dynamically) select which layer of the grid map to use for the height and which layer to use for the color, although other user-input-dependent features like enabling/disabling grid lines would be nice as well.

**Reply by banisadr** [reply]
@rg-gravis FYI, we just released [support for Grid Map](https://docs.foxglove.dev/changelog#%EF%B8%8F-support-for-grid-maps-in-3d) in the 2.27.0 release!

**Comment by lucafei** [comment]
a request for gridmap  support in 3d panel via protobuf. By now it's possible to visualize gridmap message from ros in foxglove, but for no-ros user i'd better to visualize gridmap via protobuf and this would be nice as well

**Reply by banisadr** [reply]
@lucafei Grid Map support is now[ available as of 2.27.0](https://docs.foxglove.dev/changelog#%EF%B8%8F-support-for-grid-maps-in-3d)
